### ISSUE TITLE
Provide children contacts of current contact to contact-summary calculation

### ISSUE THREAD

**Issue by jkuester** [medic/cht-core#9104]
**Is your feature request related to a problem? Please describe.**
When calculating the contact summary for a place, it would be useful to be able to include data about the child contacts of the place in the calculation.  (E.g. [to show a warning](https://forum.communityhealthtoolkit.org/t/expose-child-count-to-contact-summary/3493) if the number of contacts does not match what is expected).  Currently, there is no way to access any information about those child contacts.

**Describe the solution you'd like**
In webapp, when we have a contact in context (e.g. we are viewing that contact's profile page), we have already done the heavy lifting to [look up the contact's children](https://github.com/medic/cht-core/blob/6fb510359c95eaaf1c751451046490268d35d2b8/webapp/src/ts/effects/contacts.effects.ts#L82).  (This involves finding which contact types are valid children of the current contact and then querying `medic-client/contacts_by_parent` for each of the types.)  We only calculate the contact-summary when we have a contact in context, so I do not think adding support for this functionality would negatively impact performance (except perhaps when loading a form for a _task_.  I believe we are currently calculating a the contact-summary in that case, but we might not have the children loaded already...)

**Describe alternatives you've considered**
We could rework our contact hierarchy model to be doubly linked so contacts contain the _ids both for their parent and their children.  Seems like that would cause more trouble than it would resolve....

**Comment by jkuester** [comment]
@garethbowen after taking a deep dive into the code here and what it would take to make sure we have the contact children everywhere that we are currently calculating the contact summary, I have confirmed that we would need to add an additional call to `medic-client/contacts_by_parent` _each time we open an app form_ either from a contact profile page or from a task.  This is the only way we could guarantee we have up-to-date data about the contact's children.  

Let me know if you think this qualifies as "much of a lift"!  On one hand, the view query should be pretty straightforward and efficient. On the other hand, I know that we are already pretty sensitive about the form opening times and this will def not improve that. But, on the other hand, we would be able to do the query simultaneously with these other three lookups that we are already doing for the contact:

- getting all the contact's reports
- getting the contact's lineage
- getting the contact's target docs

These 4 things would all be happening at the same time and it seems unlikely the new query would be any slower than `medic-client/reports_by_subject`. 🤔 🤷

**Comment by dianabarsan** [comment]
> happening at the same time

Unfortunately, the indexeddb adapter is single-threaded and queues all requests. 

>  pretty sensitive about the form opening times

I agree this is not a heavy duty query. There are no free features.

**Comment by garethbowen** [comment]
The performance should be ok for families (though let's test to make sure). We will need to think about if we also need to do this looking further up the tree where there could be hundreds of children.

**Comment by jkuester** [comment]
> We will need to think about if we also need to do this looking further up the tree where there could be hundreds of children.

The more I think about this, the sketchier it gets...  On one hand, these are the same lookups that we do when loading a contact's profile page. So loading times might not be _great_, but it should not be prohibitively bad.  _However_ one big caveat here is that I believe the contact children for the profile _are paged lookups_ (so we are only getting 50 at a time or something).  Where as I feel like we really need to pass all or nothing here when it comes to the children.  (I don't want to run into the same issue we did with the `reports` where only a subset was available to the contact summary calculation....)  This seems to change the equation such that the ceiling for how bad the loading time might be for contacts with lots of children is much higher.  

IDK @garethbowen it almost feels like it would be better to solve the problem of running async code in the contact-summary logic. Then we would not have to pre-load children for every contact, but the contact summary config code would have access to the cht-datasource and could load the children if needed. 🤔 I have logged https://github.com/medic/cht-core/issues/9623 to give this possibility a closer look!

**Comment by garethbowen** [comment]
Yeah I think you're right. Otherwise we end up doing more and more preloading before calling the function without even knowing if it's going to be used.

**Comment by jkuester** [comment]
Just removing the release milestone for now.  I believe this issue will be best address by https://github.com/medic/cht-core/issues/9623

**Comment by Abhishek9639** [comment]
Hi @jkuester, 
I'd love to work on this if it's still available.

I've read through the whole thread carefully. I understand the core challenge. When calculating a contact summary for a place, there's currently no access to child contact data, which makes it impossible to do things like warn if the number of children doesn't match what's expected. I also noted the performance concern around indexeddb being single-threaded, and that the team leaned towards having the contact-summary config code request children lazily (via #9623) rather than pre-loading them unconditionally.

My plan would be to approach it in two parts:

1. First, understand the current state of #9623. If async contact-summary support is already partially in place, I'd build on top of that so the contact-summary config code can call a provided helper (e.g. getChildren()) only when it actually needs it, avoiding the cost for contacts that don't use it
2. If #9623 isn't ready yet, implement a simpler initial version where children are fetched via medic-client/contacts_by_parent at the same time as the other parallel lookups (reports, lineage, target docs) when opening a form from a contact profile page and add a performance test to confirm it's not adding meaningful overhead

Could you please assign this to me? I'll start by reviewing #9623 and update you with my findings before writing any code. Thanks

**Comment by jkuester** [comment]
Thanks for the interest @Abhishek9639, but I do not think there is currently anything actionable here.  😓 

As I noted [in this comment](https://github.com/medic/cht-core/pull/9393#issuecomment-2471621763) our latest thinking is that the performance impacts are probably significant enough that this should not be generic default functionality.  However, aside from https://github.com/medic/cht-core/issues/9623 (which has its own problems), we have not come up with a good way to mitigate these concerns.

**Comment by Abhishek9639** [comment]
Thanks for the clarification. That makes sense regarding the performance concerns.

I’d still be interested in exploring this area maybe trying out some approaches or prototypes to see if there’s a viable mitigation. Would it be okay if I experiment with this and share findings here?

Also, if there are any related issues or areas where contributions would be more useful right now, I’d be happy to take a look.

**Comment by jkuester** [comment]
👍 If you are interested in doing a bit of digging, can you actually have a look at https://github.com/medic/cht-core/issues/9623?  

The main challenge I know of in that case is going to be figuring out how to decide to [bust the cache](https://github.com/medic/cht-core/blob/d5ceadec239e1c5e4fc9603dd274fdb30584a3a2/webapp/src/ts/services/user-contact-summary.service.ts#L44) for the user's contact-summary. I have not been able to think of any good ways to approach that yet... 😓
